### ISSUE TITLE
Generic Preload Web Spec

### ISSUE THREAD

**Issue by guybedford** [systemjs/systemjs#1953]
One of the important features from SystemJS 1 was the [`depCache` configuration](https://github.com/systemjs/systemjs/blob/0.21/docs/config-api.md#depcache). This is the solution to the waterfall problem of dependency loading, that supports dynamic importing and deferred lazy loading.

This configuration allows us to indicate the module dependencies for every module in the graph, and then fetch all those dependency modules in parallel, with caching support, but only at the time they are known to be needed (we don't want to preload everything upfront due to dynamic imports).

In the same way that the resolution manifest has moved into import maps, we need to work out what the feature replacement for depCache is in this new model.

For a full write up on this problem, see the gist at https://gist.github.com/guybedford/1d36ad3040603b9dc7026df5ea80607f.

One proposal would be a dynamic **resource dependency manifest** for HTML, which could in the same way store a list of all URL resources, and their direct associated dependency URL resources.

A nice thing about doing this at the HTML level would be getting the ability to handle other asset types, as well as the ability to possibly indicate a _priority_. For example some resources would be _might be loaded_ (such as conditionals), while others would be _definitely will be loaded_ (such as direct static dependencies).

There is still some integration between import maps and such a manifest, since import maps controls the manifest, but perhaps we could make it clear that the specifiers listed in the manifest are unresolved and should still go through any resolution mechanism.

The alternative would be to try and get this into import maps itself, but there has been some resistance to this previously (see the thread at https://github.com/WICG/import-maps/issues/21).

### Example

Consider an app that has a main homepage (`/index.html`, `/index.js`, `/index.css` in turn referencing `/home.png`), that in turn has two dynamic sections it can load `/feature1.js`, `/feature1.css` and `/feature2.js`, `/feature2.css`, where the CSS files in turn reference assets like `feature1.png` and `feature2.png`. In addition there exist shared libraries used between these modules, where feature1, feature2, and index each have some cross-over of shared libraries (the full Venn diagram!). 

We don't know ahead of time whether the user will load `/feature1.js` or `/feature2.js` or even load these features at all. So we don't want load them statically on page load.

Without any preloading a request to `feature1.js` results in a latency waterfall for:

1. First a request for `feature1.js`
2. Then `feature1.css`, `sharedlib1.js` (referenced by `feature1.js`)
3. Then `feature1.png` (referenced by `feature1.css`), `sharedlib1-dep.js1` (say referenced by `sharedlib1.js`)

That is three round trips to load a feature dynamically makes the idea of shipping small minimal JS impossible as this is unacceptable.

A preload manifest for this application might look like the following in index.html:

```html
<script type="preload-manifest" src="./manifest.json"></script>
```

where the manifest would be defined with the following:

```js
{
  "/index.css": [{
    "url": "/home.png",
    "as": "image"
  }],
  "/feature1.js": [{
    "url": "/sharedlib1.js",
    "as": "module"
  }],
  "/feature2.js": [{
    "url": "/sharedlib1.js",
    "as": "module"
  }, {
    "url": "/sharedlib2.js",
    "as": "module"
  }],
  "/feature1.css": [{
    "url": "/feature1.png",
    "as": "image"
  }],
  "/feature2.css": [{
    "url": "/feature2.png",
    "as": "style"
  }],
  "/sharedlib1.js": [{
    "url": "/sharedlib1-dep.js",
    "as": "module"
  }]
}
```

Note that the manifest only has to include the first-level dependencies for any asset. This way the browser can build up the graph.

Now when the page loads, the browser can get all the information it needs without an extra round trip:

* A request to `feature1.js` can easily trace the graph of dependencies and load `/sharedlib1.js`, `/sharedlib1-dep.js`, `/feature1.css` and `/feature1.png` in parallel.
* If any of these resources already existed in the cache they don't even need a request at all (this is the benefit over bundling, which may send unnecessary data, while this approach will never send unnecessary bytes).

Additional features for the manifest include the lazy dependency declaration for index which allows pre-emptive preloading by the browser when idle, perhaps with priorities:

```js
{
  "/index.html": [{
    "url": "/feature1.js",
    "lazy": true,
    "priority": 5
  }, {
    "url": "/feature2.js",
    "lazy": true,
    "priority": 5
  }]
}
```

We can also upgrade the manifest to support integrity metadata on the base-level asset declaration:

```js
  "/feature1.js": {
    "integrity": "sha256-....",
    "preload": [{
      "url": "/sharedlib1.js",
      "as": "module"
    }]
  }
```

The manifest can support any standard preloading attributes.

**Comment by guybedford** [comment]
Summary: We need to either work on a separate WICG spec for preloading manifests to solve the waterfall problem, or we need to integrate such a spec into import maps itself.

While web package will provide an alternative approach to these problems, that doesn't mean that we shouldn't prioritise other work based on different primitives in this space.

**Comment by joeldenning** [comment]
This is a big topic. I am struggling to understand [web package](https://github.com/WICG/webpackage), though. The closest I've come to understanding it is that it's related to trusting servers when you don't have HTTPS to do that for you? I am likely vastly misunderstanding it, but how would that help with preloading dependencies of dependencies?

**Comment by justinfagnani** [comment]
@joeldenning there are two main parts to web packaging: signed exchanges and bundling. Bundling  is what avoids the waterfall problem.

**Comment by guybedford** [comment]
Unfortunately it is very hard to get awareness of this problem at a standards level still, so I think we're likely 12 months away from a generic preload spec discussion :(

**Comment by guybedford** [comment]
Note such a spec can also include subresource integrity for modules - since integrity can be associated with preload already.

**Comment by guybedford** [comment]
I've fleshed out a full example section of a preload manifest in action in the description of this issue.

**Comment by guybedford** [comment]
Depcache has now been implemented in https://github.com/systemjs/systemjs/pull/2134.
